### ISSUE TITLE
`one_to_many = return_all` output design

### ISSUE THREAD

**Issue by fontikar** [traitecoevo/APCalign#120]
`one_to_many = return_all` in `create_taxonomic_update_lookup` is returning many rows of alignments because `taxonIDClean` for the same `original_name` and `align_name` is unique

- How: Identify the combination of columns where running distinct() will return the correct number of alignments and updated_name for a taxa. Note we don't care about the number of times the an old name changes. A good edge case: a species that is broken up into many and how the output changes for all levels of `one_to_many`
- Acceptance criteria/ What is considered "done" here: Returning rows where orignal_name is aligned and changed, and not every time original_name is flagged in APC.

**Comment by wcornwell** [comment]
There is a distinct call here:
https://github.com/traitecoevo/APCalign/blob/c1001e967922279aac1493be1ae11dc6261884d4/R/update_taxonomy.R#L185
but the issue is that `taxonIDClean` has multiple entries for a given `accepted_name` which is confusing.  

Even more confusing this isn't true for all taxonomic splits.  Only sometimes and I don't know why.  So we're deep into an edge case here.  

I'd propose rationalizing `taxonIDClean, `acceptedNameUsageID`, `ccAtrributionIRI` higher up in the workflow and this will go away?  This may also be fixed as a side effect by @ehwenk working on #121, so I'll postpone this until after #121.  

Code to observe the problem is here:

```
aa<-create_taxonomic_update_lookup("Hibertia sericea",resources=resources,full=TRUE)
nrow(aa)
length(unique(aa$canonical_name))
```
potential test:
```
expect_equal(nrow(aa),length(unique(aa$canonical_name)))
```

**Comment by fontikar** [comment]
Actually this has always been a behaviour of the function. I went back to an older commit 812c1f7  and ran code to observe "problem" and there is was. Maybe we can just close this for now then...? or add back to backlog and remove from current sprint.

**Comment by wcornwell** [comment]
let's wait and reassess after #121

**Comment by ehwenk** [comment]
Some thoughts after reworking the output for update_taxonomy today. It took me a long time to recreate Will’s original output but once I did it made sense that this was the correct outcome. First, merging in the multiples from splits needs to occur before synonyms (and similar) are updated because the splits have often (not always) occurred on an outdated name. Second the canonical name and accepted name might both be identical across rows, but the taxon_ID really is unique.

**Comment by wcornwell** [comment]
> First, merging in the multiples from splits needs to occur before synonyms (and similar) are updated because the splits have often (not always) occurred on an outdated name. 

arrggg, hadn't thought about this. what's an example species?

> Second the canonical name and accepted name might both be identical across rows, but the taxon_ID really is unique.

Is this an authorship update?

**Comment by ehwenk** [comment]
@wcornwell Your workflow had already merged in the multiples from splits before the synonyms were updated - I broke your output when I changed that. But I've reverted. `Hibbertia sericea` is a good example.

I think the many taxon_ID's for a single `accepted_name_usage_ID` are about authorship. Only one of these should be retained.

**Comment by fontikar** [comment]
I think a lengthier discussion is needed for this one!

**Comment by ehwenk** [comment]
@wcornwell 

Is it correct that these are the canonical_names that have been split?
```
split_taxa_canonical <- 
  resources$APC %>%
  dplyr::filter(pro_parte == TRUE) %>%
  dplyr::distinct(canonical_name)
```

And therefore, these are the rows from APC that need to be merged in to include all splits?

```
split_taxa_all <- 
  resources$APC %>%
  dplyr::filter(pro_parte == TRUE | taxonomic_status == "accepted") %>%
  dplyr::filter(canonical_name %in% split_taxa_canonical$canonical_name) %>%
  dplyr::select(canonical_name, accepted_name_usage, accepted_name_usage_ID, taxon_ID, taxonomic_status) %>%
  dplyr::arrange(taxonomic_status)
```

**Comment by ehwenk** [comment]
to elaborate...
```
# with splits, a single canonical name, matches to multiple accepted_name_usage's; this is a one-x-to-many-y join
# "Acacia aneura" is a good example, and has been split into 3

split_taxa_canonical <- 
  resources$APC %>%
  dplyr::filter(pro_parte == TRUE) %>%
  dplyr::distinct(canonical_name)


with_split_taxa <- 
    resources$APC %>%
    dplyr::filter(pro_parte == TRUE | taxonomic_status == "accepted") %>% # this is the filter that needs to be applied for what names gets joined in if "return_all" is true
    dplyr::filter(canonical_name %in% split_taxa_canonical$canonical_name) %>% ## just for looking at rows during testing, not line to reuse, because all rows of data
    dplyr::select(canonical_name, accepted_name_usage_ID, taxon_ID, taxonomic_status, pro_parte) %>% ## just for looking at columns during testing, not line to reuse, because need more columns
    dplyr::mutate(
      alternate_accepted_name = ifelse(
        pro_parte == TRUE, 
        resources$'APC list (accepted)'$canonical_name[match(accepted_name_usage_ID, resources$'APC list (accepted)'$accepted_name_usage_ID)],
        NA
      )
    ) %>%
    dplyr::arrange(taxonomic_status) ### replace with proper function with `my_order`
  
  if (taxonomic_splits == "most_likely_species") {
    with_split_taxa <-
      with_split_taxa %>%
      dplyr::mutate(
        alternative_accepted_name_tmp = ifelse(is.na(alternate_accepted_name), NA, paste0(alternate_accepted_name, " (", taxonomic_status, ")"))
      ) %>%
      dplyr::group_by(canonical_name) %>%
      dplyr::arrange(taxonomic_status) %>% ## replace with proper function with `my_order`
      dplyr::mutate(alternative_accepted_names = 
                    alternative_accepted_name_tmp %>%
                    unique() %>% 
                    subset(., taxonomic_status != "accepted") %>% ## not as simple as this. See canonical_name "Acacia aneura var. conifera" which is a synonym, and has been split into "Acacia aneura" and "Acacia aptaneura". 
                      #How do you choose a most likely species?
                    paste0(collapse = " | ") %>%
                    dplyr::na_if("")
      ) %>%
      dplyr::ungroup() %>%
      dplyr::mutate(
        alternative_accepted_names = ifelse(taxonomic_status != "accepted" & canonical_name %in% resources$'APC list (accepted)'$canonical_name, NA, alternative_accepted_names),
        alternative_accepted_names = stringr::str_replace_all(alternative_accepted_names, "\\ \\|\\ NA", "")
      ) #%>% 
      #dplyr::select(-alternative_accepted_name_tmp, -alternative_accepted_name) # leaving in for now for testing 
  }


#this would become the start of the function for APC species in `update_taxonomy`... but it requires creating this completely new APC tibble - the above code

data %>%
  dplyr::left_join
    by = "aligned_name",
    with_split_taxa %>% rename(aligned_name = canonical_name)  
  ) %>% ...
```

**Comment by ehwenk** [comment]
Another part of this, is do we want duplicate entries for distinct(canonical_name, alternate_accepted_name, taxonomic_status) combinations? Each row has a unique taxon_ID, but everything else is identical. We were talking about these the other day, but we didn't really reach a conclusion. Same scientific_name, same scientific_name_ID, etc. Just different "instances" of the name that have been recorded. Will's example in tests includes the duplicates.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Lizzy/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Lizzy/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



aligned_name | taxon_ID_aligned | taxonomic_status_aligned | accepted_name_usage_ID | canonical_name | scientific_name_authorship
-- | -- | -- | -- | -- | --
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912933 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912932 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912942 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912937 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912936 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912934 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912935 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912938 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/912940 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2899370 | Hibbertia crinita | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916213 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2913496 | Hibbertia incana | (Lindl.) Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916214 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2913496 | Hibbertia incana | (Lindl.) Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916215 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2913496 | Hibbertia incana | (Lindl.) Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916216 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2913496 | Hibbertia incana | (Lindl.) Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916217 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2913496 | Hibbertia incana | (Lindl.) Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916212 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2913496 | Hibbertia incana | (Lindl.) Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916252 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2891985 | Hibbertia paeninsularis | J.M.Black
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916259 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2911420 | Hibbertia patens | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916257 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2911420 | Hibbertia patens | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916255 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2911420 | Hibbertia patens | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916270 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2907488 | Hibbertia platyphylla | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916271 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2907488 | Hibbertia platyphylla | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916287 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2892290 | Hibbertia platyphylla subsp. major | (J.M.Black) Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916281 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2896117 | Hibbertia platyphylla subsp. platyphylla | NA
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916280 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2896117 | Hibbertia platyphylla subsp. platyphylla | NA
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916434 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2900262 | Hibbertia puberula | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916432 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2900262 | Hibbertia puberula | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916433 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2900262 | Hibbertia puberula | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/node/apni/2895583 | accepted | https://id.biodiversity.org.au/node/apni/2895583 | Hibbertia sericea | (R.Br. ex DC.) Benth.
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916525 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2901708 | Hibbertia simulans | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916527 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2901708 | Hibbertia simulans | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916526 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2901708 | Hibbertia simulans | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916524 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2901708 | Hibbertia simulans | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916556 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2910589 | Hibbertia superans | Toelken
Hibbertia sericea | https://id.biodiversity.org.au/instance/apni/916566 | pro parte misapplied | https://id.biodiversity.org.au/node/apni/2915755 | Hibbertia villifera | Tepper ex Toelken



</body>

</html>

**Comment by wcornwell** [comment]
@ehwenk this looks like great progress.  I agree that `pro_parte == TRUE` is the right way to get the splits--my reading of the APC design is that this column is set up specifically for that purpose.  (I had one small query about splits that were subsequently lumped again (which happens a lot) but my guess is that they are making pro_parte=FALSE for that case, which is good for our use case.)  

> do we want duplicate entries for distinct(canonical_name, alternate_accepted_name, taxonomic_status) combinations?

No, I think we want to get rid of them. As in the Hibbertia case above.  This isn't too hard, but you're better placed than me to figure out where to do this in the workflow at the moment. My current understanding is that we should care about distinct APNI "nodes" not distinct APNI "instances".

**Comment by ehwenk** [comment]
> I agree that `pro_parte == TRUE` is the right way to get the splits--my reading of the APC design is that this column is set up specifically for that purpose.

If we agree on this, it makes determining what names to use in splits much easier!

> No, I think we want to get rid of them. As in the Hibbertia case above. This isn't too hard, but you're better placed than me to figure out where to do this in the workflow at the moment. 

I'm also in favour of removing duplicates and I know where to remove them in the workflow. I was worrying about which taxon_ID to retain, but realise it doesn't matter, because `taxon_id` will become the `accepted_name_usage_id`. We won't be maintaining a taxon_ID_aligned to the end anymore.

**Comment by ehwenk** [comment]
@wcornwell - I'm struggling here, trying to work out if I "merge in splits" before or after taxonomic updates. Using you favourite species, `Acacia aneura`, for those taxa that are taxonomic synonyms of `Acacia aneura`, are they also subject to the split ambiguity, or only if the actual `original_name` is `Acacia aneura`. This gets trickier for a few other names, where there is a split, but the `original_name` itself isn't `accepted`.

For instance `Abutilon muticum` is both pro parte misapplied and a taxonomic synonym - I can probably come up with an algorithm to just update those (few) names that are both `pro parte` and not `accepted`, but want to make sure this is what we want.

![image](https://github.com/traitecoevo/APCalign/assets/8701562/8376c5d3-b30c-4dbf-8a4d-a260f68a5c53)

**Comment by wcornwell** [comment]
> are they also subject to the split ambiguity, or only if the actual original_name is Acacia aneura

The second option.  Acacia minyura is a not subject to split ambiguity since it was a newly created name.  (It might get re-lumped in the future, but that's a problem for another day.)

> For instance Abutilon muticum is both pro parte misapplied and a taxonomic synonym 

Abutilon muticum looks like incompletely done taxonomy to me.  Current behavior isn't terrible, given we're working with incomplete information: 
<img width="686" alt="Screenshot 2023-08-24 at 9 05 44 am" src="https://github.com/traitecoevo/APCalign/assets/5534161/b1b80424-3e18-464f-9422-4edc66c56b4f">

**Comment by ehwenk** [comment]
Closed with commit ac799c3
